[postfilter-impl-2] Introduce a defaultpreemption PostFilter plugin

Huang-Wei opened the pull request:
**What type of PR is this?**

/kind feature
/sig scheduling

**What this PR does / why we need it**:

This PR firstly makes `g.Preempt()` stateless, and then introduced a defaultpreemption plugin (not activated) to call `Preempt()` directly.

It corresponds to the second part of @ahg-g's #91426 (comment)

**Which issue(s) this PR fixes**:

Part of #91038.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

ahg-g commented:
Wonderful, no comments on this one! Once https://github.com/kubernetes/kubernetes/pull/92009 is merged we can get this one in as well.

ahg-g commented:
actually , lets pass in FrameworkHandle instead of the whole profile in this PR so that https://github.com/kubernetes/kubernetes/pull/92049 only adds the plugin.

mlmhl commented on pkg/scheduler/core/generic_scheduler.go:
```diff
@@ -281,50 +278,64 @@ func (g *genericScheduler) Preempt(ctx context.Context, prof *profile.Profile, s
 		}
 	}
-	nodeNameToVictims, err := selectNodesForPreemption(ctx, prof, g.podNominator, state, pod, potentialNodes, pdbs)
+	nodeNameToVictims, err := selectNodesForPreemption(ctx, fh.PreemptHandle(), fh.PreemptHandle(), state, pod, potentialNodes, pdbs)
```
Could we make `selectNodesForPreemption` as public method so that user's custom preemption plugins can reuse this method?

Huang-Wei replied:
It will be made public in a followup PR. This PR just keep the old interfaces as is.

mlmhl replied:
Got it, thanks!

Huang-Wei commented:
@ahg-g this is ready for review. (just review the 2nd commit)

/hold

Huang-Wei commented:
/retest

ahg-g commented on pkg/scheduler/core/generic_scheduler.go:
```diff
@@ -286,21 +282,22 @@ func (g *genericScheduler) Preempt(ctx context.Context, prof *profile.Profile, s
 			return "", err
 		}
 	}
-	nodeNameToVictims, err := selectNodesForPreemption(ctx, prof, g.podNominator, state, pod, potentialNodes, pdbs)
```
Can we remove podNominator now from generic scheduler?

Huang-Wei replied:
It's better to wait until all preemption functions get moved to preemption plugin, and then I can (1) revisit each function's signature in defaultpreemption package, and (2) genericScheduler should be easy by then to be cleaned up.

ahg-g commented on pkg/scheduler/core/generic_scheduler.go:
```diff
@@ -286,21 +282,22 @@ func (g *genericScheduler) Preempt(ctx context.Context, prof *profile.Profile, s
 		klog.Infof("%v potential nodes for preemption, first %v are: %v", len(potentialNodes), len(sample), sample)
 	}
 	var pdbs []*policy.PodDisruptionBudget
-	if g.pdbLister != nil {
```
Can pdbLister now be removed from genericScheduler?

Huang-Wei replied:
Yes, let me remove it.

ahg-g commented:
/retest
/lgtm

Huang-Wei commented:
/retest
/hold cancel

Huang-Wei commented:
/hold

Huang-Wei commented:
I forgot to remove the parameter in NewGenericScheduler(), the latest change covers it.

@ahg-g needs another /lgtm.

Huang-Wei commented:
/hold cancel

ahg-g commented on pkg/scheduler/factory.go:
```diff
@@ -192,7 +192,6 @@ func (c *Configurator) create() (*Scheduler, error) {
 		c.nodeInfoSnapshot,
 		extenders,
 		c.informerFactory.Core().V1().PersistentVolumeClaims().Lister(),
-		GetPodDisruptionBudgetLister(c.informerFactory),
```
we need to remove the function itself and check the feature flag and skip evaluating pdb if not set, we can do that as a followup.

liggitt replied:
it needs to be done in this PR. As is, this PR fails the GA-only conformance e2e test

ahg-g reviewed (commented):
/lgtm

ahg-g commented:
/lgtm cancel

preemption integration tests are failing

Huang-Wei commented:
> preemption integration tests are failing

The removed but useless pdbLister happens to trigger the initilization of pdbLister. So we have to explicitly trigger the pdbLister initilization. The 2nd commit did that, but I'm not sure it's conventional.

ahg-g commented on pkg/scheduler/eventhandlers.go:
```diff
@@ -466,6 +466,8 @@ func addAllEventHandlers(
 		},
 	)
+
+	informerFactory.Policy().V1beta1().PodDisruptionBudgets().Lister()
```
@liggitt how do we make sure listers are properly initialized? It seems that we have to do something like this at scheduler initialization, or is the pattern that we have to pass a lister around instead of the sharedInformer instance itself?

ahg-g replied:
@Huang-Wei what we can do is store a reference to the lister at the preemption plugin initialization.

Huang-Wei replied:
> @Huang-Wei what we can do is store a reference to the lister at the preemption plugin initialization.

That's doable, although it sounds a bit tricky as it can be retrieved from the SharedInformerFactory. Let's first check with api experts to see the best practice here.

liggitt replied:
you cannot assume the PDB API is available (it is a beta API that can be disabled). this needs to tolerate it not being served and not block/fail startup

Huang-Wei replied:
I see. Now it's being wrapped in a featuregate checking block.

Huang-Wei replied:
So @liggitt, back to the question, can we put a bare `informerFactory....Lister()` to trigger the lister initialization? If not, is there a better way?

(the context is we don't need a pdbInformer to watch it; we just use the pdbLister to get the pdb list from the cache storage)

liggitt replied:
> can we put a bare informerFactory....Lister() to trigger the lister initialization

Sure. Doing so once (only if the feature is enabled) and saving the result is what we typically do. That's what the code did previously.

ahg-g commented:
> > preemption integration tests are failing
> 
> The removed (useless) pdbLister happens to trigger the initilization of pdbLister. So we have to explicitly trigger the pdbLister initilization. The 2nd commit did that, but I'm not sure it's conventional.

I think what we need to do is to call this at PostFilter initialization.

Huang-Wei committed:
Introduce a defaultpreemption PostFilter plugin

- Add a defaultpreemption PostFilter plugin
- Make g.Preempt() stateless
    - make g.Preempt() stateless
    - make g.getLowerPriorityNominatedPods() stateless
    - make g.processPreemptionWithExtenders() stateless

ahg-g commented:
/lgtm

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

dims commented:
/retest

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

Huang-Wei commented:
/hold cancel

Huang-Wei commented:
/retest

Huang-Wei commented:
Don't know why pull-kubernetes-conformance-kind-ga-only-parallel keeps failing. Trying to do a no-op force push...

Huang-Wei commented:
/retest

liggitt commented on pkg/scheduler/core/generic_scheduler.go:
```diff
@@ -286,21 +280,22 @@ func (g *genericScheduler) Preempt(ctx context.Context, prof *profile.Profile, s
 	var pdbs []*policy.PodDisruptionBudget
-	if g.pdbLister != nil {
-		pdbs, err = g.pdbLister.List(labels.Everything())
+	pdbLister := fh.SharedInformerFactory().Policy().V1beta1().PodDisruptionBudgets().Lister()
```
Do not assume PDBs are available. The API might not be enabled on a GA-only cluster

liggitt replied:
Make sure the feature is enabled (maybe like GetPodDisruptionBudgetLister did?) before accessing the lister like this, otherwise the scheduler will list a non-existent PDB API every second and get a 404 error forever.

Huang-Wei replied:
Ah, thanks. I did the featuregate check in a followup PR, but not this one. Let me update.

Huang-Wei replied:
This explains why the followup (#92049) passed CI, but not this one.

Huang-Wei replied:
GetPodDisruptionBudgetLister cannot be used in pkg/scheduler/core due to cycling dependency, so I put featuregate check here.

Huang-Wei commented:
/retest

ahg-g commented on pkg/scheduler/core/generic_scheduler.go:
```diff
@@ -286,21 +282,21 @@ func (g *genericScheduler) Preempt(ctx context.Context, prof *profile.Profile, s
 	var pdbs []*policy.PodDisruptionBudget
-	if g.pdbLister != nil {
-		pdbs, err = g.pdbLister.List(labels.Everything())
+	if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.PodDisruptionBudget) {
+		pdbs, err = fh.SharedInformerFactory().Policy().V1beta1().PodDisruptionBudgets().Lister().List(labels.Everything())
 		if err != nil {
 			return "", err
 		}
 	}
```
lets wrap this in a function that we move later to the postfilter plugin: 

```
func getPodDisruptionBudgets() ([]*policy.PodDisruptionBudget, error) {
  if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.PodDisruptionBudget) {
    return fh.SharedInformerFactory().Policy().V1beta1().PodDisruptionBudgets().Lister().List(labels.Everything())
}
 return nil, nil
}
```

Huang-Wei replied:
Updated.

ahg-g replied:
Thanks, when we move this in the follow up PR, we should get the lister once at postfilter instantiation and avoid constantly checking the feature flag.

Huang-Wei replied:
yup, currently the followup PR already honors featuregate, but using a different style. I will rebase changes from this PR.

Huang-Wei commented:
/retest

ahg-g commented:
/lgtm

Huang-Wei commented:
/retest
